### PR TITLE
XL/fs: ListMultipartUploads should list only requested entries.

### LINKED ISSUE #1665: list-incomplete-uploads returns more than max-keys entries
initiate 2000 incomplete uploads:

```
        {
                for i := 0; i < 2000; i++ {
                        if _, err := client.initiateMultipartUpload(bucketName, fmt.Sprintf("%s.%d", objectName, i), ""); err != nil {
                                log.Fatal(err)
                        }
                }
        }
```

`testobject0.1898` is the 1000th entry, create one more uploadID for it:

```

        if res, err := client.initiateMultipartUpload(bucketName, fmt.Sprintf("%s.%d", objectName, 1898), ""); err != nil {
                log.Fatal(err)
        } else {
                fmt.Println(res)
        }

```

Now do `mc ls -I myminio/testbucket/`

and print the number of entries in the result of list-incomplete-uploads:

```
krishna@escape:~/gocode/src/github.com/minio/minio$ git diffdiff --git a/object-common-multipart.go b/object-common-multipart.go
index 5f9a2b7..3c0e0d1 100644
--- a/object-common-multipart.go
+++ b/object-common-multipart.go
@@ -467,6 +467,7 @@ func listMultipartUploadsCommon(layer ObjectLayer, bucket, prefix, keyMarker, up
                result.NextKeyMarker = ""
                result.NextUploadIDMarker = ""
        }
+       fmt.Println(len(result.Uploads), result.NextKeyMarker, result.NextUploadIDMarker)
        return result, nil
 }

krishna@escape:~/gocode/src/github.com/minio/minio$ 
```

it prints:

```
1001 testobject0.1898 8f60f886-83f3-4e49-8c78-ce89d462687f
1000 testobject0.999 b3d9054c-8008-494f-9dd5-ef8aae885a94
0
```

i.e it returns 1001 entries when 1000 entries were requested.

### PR DESCRIPTION
**Pull request by harshavardhana**
// cc @krishnasrinivas 

Fixes #1665

### CODE REVIEW

File: object-common-multipart.go [hunk 1]
```diff
@@ -336,6 +336,7 @@ func listMetaBucketMultipartFiles(layer ObjectLayer, prefixPath string, markerPa
 				}
 				fileInfo.Name = path.Join(fi.Name, entry)
 				fileInfos = append(fileInfos, fileInfo)
+				newMaxKeys++
```
Comment (krishnasrinivas) [inline object-common-multipart.go:363]: here you need to check if `newMaxKeys == maxKeys`

**Comment by krishnasrinivas** [comment]
@harshavardhana it's a little tricky because:
`In the response, the uploads are sorted by key. If your application has initiated more than one multipart upload using the same object key, then uploads in the response are first sorted by key. Additionally, uploads are sorted in ascending order within each key by the upload initiation time.`

**Comment by krishnasrinivas** [comment]
we have the option of not conforming to S3 `Additionally, uploads are sorted in ascending order within each key by the upload initiation time` and sort it lexically and keep the code simple.

**Comment by harshavardhana** [comment]
> @harshavardhana it's a little tricky because:
> In the response, the uploads are sorted by key. If your application has initiated more than one multipart upload using the same object key, then uploads in the response are first sorted by key. Additionally, uploads are sorted in ascending order within each key by the upload initiation time.

Won't fix this in this patch - the problem here is that "listLeafEntries" are sorted as is. Not just based on keyNames, but based on "keyName/upload-id" so lexically it provides a different meaning. 

So at the top level inside ListMultipartUploads the context is already very different.  What we need to do is these leafEntries should be sorted based on 'keys' first not like 'key/upload-id'.  Then we need to use put them back in the key order in fileInfos.  Additionally keep track of duplicates so that the `initiatedTime` is used additionally to sort again. 

We can discuss, but i don't know if its a big issue.

**Comment by krishnasrinivas** [comment]
looks good

**Pull request merged**